File: typescript/src/utils/dividend-score.ts
/**
 * Dividend Safety Score.
 *
 * Assesses the probability that a company can maintain and grow its dividend
 * using multi-year financial data. The analysis covers free cash flow coverage,
 * earnings coverage, balance sheet capacity, and dividend growth track record.
 *
 * Non-payers receive a special rating rather than a scored assessment.
 *
 * Signals (dividend-payers only)
 * -------
 * 1. FCF Payout Ratio       (35%) — dividends as fraction of free cash flow
 * 2. Earnings Payout Ratio  (25%) — dividends as fraction of net income
 * 3. Balance Sheet Strength (25%) — net debt / EBITDA capacity headroom
 * 4. Dividend Growth Track  (15%) — consecutive years of non-declining dividends
 *
 * Score:  0–100
 * Rating: 'safe' (≥70), 'adequate' (45–69), 'risky' (20–44), 'danger' (<20), 'non-payer'
 */

// ── Types ──────────────────────────────────────────────────────────────────────

export interface DividendComponents {
  fcfPayoutRatio: number        // 0–1
  earningsPayoutRatio: number   // 0–1
  balanceSheetStrength: number  // 0–1
  dividendGrowthTrack: number   // 0–1
}

export interface DividendSafetyScore {
  score: number
  rating: 'safe' | 'adequate' | 'risky' | 'danger' | 'non-payer'
  components: DividendComponents
  isDividendPayer: boolean
  yearsAnalyzed: number
  evidence: string[]
  interpretation: string
}

/** One year of financial data for dividend safety analysis. */
export type AnnualDividendData = {
  dividendsPaid?: number
  operatingCashFlow?: number
  capex?: number
  netIncome?: number
  totalDebt?: number
  cash?: number
  ebit?: number
  depreciation?: number
  year?: number | string
}

// ── Helpers ────────────────────────────────────────────────────────────────────

function mean(xs: number[]): number {
  return xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : 0
}

function clamp(x: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, x))
}

function g(d: AnnualDividendData, k: keyof AnnualDividendData): number {
  const v = d[k]
  return typeof v === 'number' && isFinite(v) ? v : 0
}

// ── Payer detection ───────────────────────────────────────────────────────────

function isDividendPayer(series: AnnualDividendData[]): boolean {
  return series.some(d => {
    const div = d.dividendsPaid
    return typeof div === 'number' && isFinite(div) && div > 0
  })
}

// ── Signal 1: FCF Payout Ratio ────────────────────────────────────────────────

function scoreFcfPayout(series: AnnualDividendData[]): [number, string[]] {
  const ratios: number[] = []
  for (const d of series) {
    const div = g(d, 'dividendsPaid')
    if (div <= 0) continue
    const cfo = g(d, 'operatingCashFlow')
    const capex = g(d, 'capex')
    const fcf = cfo - capex
    if (fcf !== 0) ratios.push(div / fcf)
  }
  if (!ratios.length) return [0.50, ['FCF payout ratio: no dividend/FCF overlap (neutral score)']]

  const avg = mean(ratios)
  let score: number
  if (avg < 0.4)       score = 1.0
  else if (avg < 0.6)  score = 0.75
  else if (avg < 0.8)  score = 0.45
  else if (avg <= 1.0) score = 0.15
  else                 score = 0.0

  const quality = avg < 0.4 ? 'very well covered' : avg < 0.6 ? 'adequately covered' : avg < 0.8 ? 'tight' : avg <= 1.0 ? 'stressed' : 'uncovered by FCF'
  return [score, [
    `FCF payout ratio: mean ${(avg * 100).toFixed(0)}%  (${quality})`,
  ]]
}

// ── Signal 2: Earnings Payout Ratio ───────────────────────────────────────────

function scoreEarningsPayout(series: AnnualDividendData[]): [number, string[]] {
  const ratios: number[] = []
  for (const d of series) {
    const div = g(d, 'dividendsPaid')
    const ni = g(d, 'netIncome')
    if (div > 0 && ni > 0) ratios.push(div / ni)
  }
  if (!ratios.length) return [0.50, ['Earnings payout ratio: no dividend/earnings overlap (neutral score)']]

  const avg = mean(ratios)
  let score: number
  if (avg < 0.35)      score = 1.0
  else if (avg < 0.55) score = 0.75
  else if (avg < 0.75) score = 0.45
  else if (avg <= 1.0) score = 0.20
  else                 score = 0.0

  const quality = avg < 0.35 ? 'conservative' : avg < 0.55 ? 'moderate' : avg < 0.75 ? 'elevated' : avg <= 1.0 ? 'high' : 'exceeds earnings'
  return [score, [
    `Earnings payout ratio: mean ${(avg * 100).toFixed(0)}%  (${quality})`,
  ]]
}

// ── Signal 3: Balance Sheet Strength ──────────────────────────────────────────

function scoreBalanceSheet(series: AnnualDividendData[]): [number, string[]] {
  // Use most recent year
  const d = series[series.length - 1]!
  const totalDebt = g(d, 'totalDebt')
  const cash = g(d, 'cash')
  const netDebt = totalDebt - cash
  const ebit = g(d, 'ebit')
  const dep = g(d, 'depreciation')
  const ebitda = ebit + dep

  if (ebitda <= 0) {
    return [0.30, ['Balance sheet strength: EBITDA unavailable or negative (below-neutral score)']]
  }

  const ratio = netDebt / ebitda
  let score: number
  if (ratio < 0)       score = 1.0
  else if (ratio < 1)  score = 0.85
  else if (ratio < 2)  score = 0.65
  else if (ratio < 3)  score = 0.40
  else if (ratio < 4)  score = 0.20
  else                 score = 0.05

  const quality = ratio < 0 ? 'net cash position' : ratio < 1 ? 'very low leverage' : ratio < 2 ? 'modest leverage' : ratio < 3 ? 'moderate leverage' : ratio < 4 ? 'elevated leverage' : 'highly levered'
  return [score, [
    `Net debt / EBITDA: ${ratio.toFixed(2)}×  (${quality})`,
  ]]
}

// ── Signal 4: Dividend Growth Track ───────────────────────────────────────────

function scoreDividendGrowthTrack(series: AnnualDividendData[]): [number, string[]] {
  const divs = series.map(d => g(d, 'dividendsPaid')).filter(d => d > 0)
  if (divs.length < 2) {
    return [0.15, ['Dividend growth track: insufficient history (below-neutral score)']]
  }

  let consecutive = 0
  for (let i = 1; i < divs.length; i++) {
    if ((divs[i] ?? 0) >= (divs[i - 1] ?? 0)) {
      consecutive++
    } else {
      consecutive = 0  // reset on any cut
    }
  }

  let score: number
  if (consecutive >= 10)     score = 1.0
  else if (consecutive >= 7) score = 0.75
  else if (consecutive >= 4) score = 0.55
  else if (consecutive >= 2) score = 0.35
  else                       score = 0.15

  const label = consecutive >= 10 ? 'dividend aristocrat-level streak' : consecutive >= 7 ? 'strong track record' : consecutive >= 4 ? 'solid track record' : consecutive >= 2 ? 'limited track record' : 'recent cut or minimal history'
  return [score, [
    `Dividend growth track: ${consecutive} consecutive non-declining year(s)  (${label})`,
  ]]
}

// ── Main function ─────────────────────────────────────────────────────────────

/**
 * Compute Dividend Safety Score from a sequence of annual financial records.
 *
 * @param annualData - Array in chronological order (oldest first). Min 2 years recommended.
 * @returns DividendSafetyScore with score (0–100), rating, components, and evidence.
 *
 * @example
 *   const result = dividendSafetyScoreFromSeries([
 *     { year: 2020, dividendsPaid: 1.2e9, operatingCashFlow: 8e9, capex: 2e9,
 *       netIncome: 5e9, totalDebt: 10e9, cash: 3e9, ebit: 7e9, depreciation: 1e9 },
 *     { year: 2021, dividendsPaid: 1.3e9, operatingCashFlow: 9e9, capex: 2.2e9,
 *       netIncome: 5.5e9, totalDebt: 9e9, cash: 3.5e9, ebit: 7.8e9, depreciation: 1.1e9 },
 *   ])
 *   console.log(result.score)   // e.g. 74
 *   console.log(result.rating)  // 'safe'
 */
export function dividendSafetyScoreFromSeries(
  annualData: AnnualDividendData[],
): DividendSafetyScore {
  const payer = isDividendPayer(annualData)

  if (!payer) {
    return {
      score: 50,
      rating: 'non-payer',
      isDividendPayer: false,
      yearsAnalyzed: annualData.length,
      components: {
        fcfPayoutRatio:      0.5,
        earningsPayoutRatio: 0.5,
        balanceSheetStrength: 0.5,
        dividendGrowthTrack: 0.5,
      },
      evidence: ['No dividends detected across the provided data series'],
      interpretation: 'Score 50/100: Company does not pay a dividend — score is not meaningful',
    }
  }

  const [fcfScore, fcfEv] = scoreFcfPayout(annualData)
  const [epScore, epEv] = scoreEarningsPayout(annualData)
  const [bsScore, bsEv] = scoreBalanceSheet(annualData)
  const [dgtScore, dgtEv] = scoreDividendGrowthTrack(annualData)

  const raw = 0.35 * fcfScore + 0.25 * epScore + 0.25 * bsScore + 0.15 * dgtScore
  const score = Math.round(clamp(raw, 0, 1) * 100)

  const rating: DividendSafetyScore['rating'] =
    score >= 70 ? 'safe' :
    score >= 45 ? 'adequate' :
    score >= 20 ? 'risky' : 'danger'

  const desc: Record<Exclude<DividendSafetyScore['rating'], 'non-payer'>, string> = {
    safe:     'Dividend well-covered by cash flows with a strong balance sheet',
    adequate: 'Dividend appears sustainable but headroom is limited in some areas',
    risky:    'Coverage is strained — dividend could be at risk in a downturn',
    danger:   'Dividend likely unsustainable; cut probability is high',
  }

  return {
    score,
    rating,
    components: {
      fcfPayoutRatio:       Math.round(fcfScore  * 1e4) / 1e4,
      earningsPayoutRatio:  Math.round(epScore   * 1e4) / 1e4,
      balanceSheetStrength: Math.round(bsScore   * 1e4) / 1e4,
      dividendGrowthTrack:  Math.round(dgtScore  * 1e4) / 1e4,
    },
    isDividendPayer: true,
    yearsAnalyzed: annualData.length,
    evidence: [...fcfEv, ...epEv, ...bsEv, ...dgtEv],
    interpretation: `Score ${score}/100: ${desc[rating]}`,
  }
}

File: python/fin_ratios/utils/dividend_score.py
"""
Dividend Safety Score.

Assesses the sustainability of a company's dividend based on cash flow
coverage, earnings payout ratio, balance sheet strength, and dividend
growth track record. Non-payers receive a special 'non-payer' designation
rather than being penalised.

Score:  0–100
Rating: 'safe' (≥70), 'adequate' (45–69), 'risky' (20–44), 'danger' (<20)
Special: 'non-payer' when the company has not paid dividends in any year
         analysed (score=50, is_dividend_payer=False).

Signals
-------
1. FCF Payout Ratio     (35%) — dividends / free cash flow
2. Earnings Payout Ratio (25%) — dividends / net income
3. Balance Sheet Strength (25%) — net debt / EBITDA
4. Dividend Growth Track  (15%) — consecutive years of dividend maintenance

References
----------
Siegel, J.J. (2014) — Stocks for the Long Run (5th ed.), McGraw-Hill
Dividend Aristocrats methodology — S&P Dow Jones Indices
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any, Sequence


# ── Field registry ─────────────────────────────────────────────────────────────

_ZERO_FIELDS = frozenset(
    {
        "revenue",
        "ebit",
        "net_income",
        "total_assets",
        "total_equity",
        "total_debt",
        "cash",
        "capex",
        "depreciation",
        "interest_expense",
        "income_tax_expense",
        "operating_cash_flow",
        "dividends_paid",
        "current_assets",
        "current_liabilities",
        "ebt",
    }
)

_ALIASES: dict[str, list[str]] = {
    "ebit": ["operating_income"],
    "total_debt": ["long_term_debt"],
    "depreciation": ["depreciation_amortization"],
    "dividends_paid": ["dividends", "dividend_payments", "dividends_and_other_cash_distributions"],
    "operating_cash_flow": [
        "cash_from_operations",
        "cfo",
        "net_cash_from_operating_activities",
        "operating_activities",
    ],
}


class _Acc:
    """Uniform attribute access for dicts, dataclasses, and arbitrary objects."""

    def __init__(self, src: Any) -> None:
        object.__setattr__(self, "_src", src)

    def _raw(self, name: str) -> Any:
        src = object.__getattribute__(self, "_src")
        if isinstance(src, dict):
            return src.get(name)
        return getattr(src, name, None)

    def __getattr__(self, name: str) -> float:
        val = self._raw(name)
        if val is not None and not (isinstance(val, float) and math.isnan(val)):
            return float(val)
        for alias in _ALIASES.get(name, []):
            val = self._raw(alias)
            if val is not None and not (isinstance(val, float) and math.isnan(val)):
                return float(val)
        if name == "ebt":
            ebit = self._raw("ebit") or self._raw("operating_income") or 0.0
            interest = self._raw("interest_expense") or 0.0
            if ebit or interest:
                return float(ebit) - float(interest)
        if name in _ZERO_FIELDS:
            return 0.0
        raise AttributeError(name)


# ── Statistical helpers ────────────────────────────────────────────────────────


def _mean(xs: list[float]) -> float:
    return sum(xs) / len(xs) if xs else 0.0


def _clamp(x: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, x))


# ── FCF helper ─────────────────────────────────────────────────────────────────


def _year_fcf(d: _Acc) -> float:
    """FCF = OCF - capex, falling back to NOPAT-based if OCF unavailable."""
    ocf = d.operating_cash_flow
    capex = d.capex
    if ocf > 0:
        return ocf - capex
    # NOPAT-based fallback
    ebit = d.ebit
    if ebit <= 0:
        return 0.0
    ebt = d.ebt
    tax = d.income_tax_expense
    tax_rate = _clamp(tax / ebt, 0.0, 0.50) if ebt > 0 and tax > 0 else 0.21
    nopat = ebit * (1.0 - tax_rate)
    return nopat + d.depreciation - capex


# ── Signal 1: FCF Payout Ratio ────────────────────────────────────────────────


def _score_fcf_payout(accs: list[_Acc]) -> tuple[float, list[str]]:
    ratios: list[float] = []
    for d in accs:
        div = d.dividends_paid
        if div <= 0:
            continue
        fcf = _year_fcf(d)
        if fcf <= 0:
            ratios.append(2.0)  # penalised: paying dividends from negative FCF
        else:
            ratios.append(div / fcf)

    if not ratios:
        return 0.50, ["FCF payout ratio: no dividend payments found (neutral score)"]

    mean_ratio = _mean(ratios)
    if mean_ratio < 0.40:
        score = 1.0
    elif mean_ratio < 0.60:
        score = 0.75
    elif mean_ratio < 0.80:
        score = 0.45
    elif mean_ratio <= 1.00:
        score = 0.15
    else:
        score = 0.0

    quality = (
        "very safe"
        if mean_ratio < 0.40
        else "safe"
        if mean_ratio < 0.60
        else "moderate"
        if mean_ratio < 0.80
        else "stretched"
        if mean_ratio <= 1.00
        else "unsustainable"
    )
    return score, [
        f"FCF payout ratio: mean {mean_ratio * 100:.0f}%  [{quality}]",
    ]


# ── Signal 2: Earnings Payout Ratio ───────────────────────────────────────────


def _score_earnings_payout(accs: list[_Acc]) -> tuple[float, list[str]]:
    ratios: list[float] = []
    for d in accs:
        div = d.dividends_paid
        ni = d.net_income
        if div <= 0:
            continue
        if ni <= 0:
            ratios.append(2.0)
        else:
            ratios.append(div / ni)

    if not ratios:
        return 0.50, ["Earnings payout ratio: no dividend payments found (neutral score)"]

    mean_ratio = _mean(ratios)
    if mean_ratio < 0.35:
        score = 1.0
    elif mean_ratio < 0.55:
        score = 0.75
    elif mean_ratio < 0.75:
        score = 0.45
    elif mean_ratio <= 1.00:
        score = 0.20
    else:
        score = 0.0

    quality = (
        "conservative"
        if mean_ratio < 0.35
        else "moderate"
        if mean_ratio < 0.55
        else "elevated"
        if mean_ratio < 0.75
        else "high"
        if mean_ratio <= 1.00
        else "exceeds earnings"
    )
    return score, [
        f"Earnings payout ratio: mean {mean_ratio * 100:.0f}%  [{quality}]",
    ]


# ── Signal 3: Balance Sheet Strength ──────────────────────────────────────────


def _score_balance_sheet(accs: list[_Acc]) -> tuple[float, list[str]]:
    d = accs[-1]  # most recent year
    total_debt = d.total_debt
    cash = d.cash
    net_debt = total_debt - cash

    ebit = d.ebit
    depreciation = d.depreciation
    ebitda = ebit + depreciation if ebit > 0 else 0.0

    if ebitda <= 0:
        # Fall back: use net income + tax + interest + depreciation approximation
        ni = d.net_income
        tax = d.income_tax_expense
        interest = d.interest_expense
        ebitda = ni + tax + interest + depreciation
        if ebitda <= 0:
            return 0.30, ["Balance sheet strength: unable to compute EBITDA (conservative score)"]

    nd_ebitda = net_debt / ebitda

    if nd_ebitda < 0:
        score = 1.0
        quality = "net cash position"
    elif nd_ebitda <= 1.0:
        score = 0.85
        quality = "minimal leverage"
    elif nd_ebitda <= 2.0:
        score = 0.65
        quality = "modest leverage"
    elif nd_ebitda <= 3.0:
        score = 0.40
        quality = "moderate leverage"
    elif nd_ebitda <= 4.0:
        score = 0.20
        quality = "high leverage"
    else:
        score = 0.05
        quality = "very high leverage"

    return score, [
        f"Net debt / EBITDA: {nd_ebitda:.1f}x  [{quality}]  "
        f"(net debt {net_debt / 1e9:.1f}B, EBITDA {ebitda / 1e9:.1f}B)",
    ]


# ── Signal 4: Dividend Growth Track ───────────────────────────────────────────


def _score_dividend_growth_track(accs: list[_Acc]) -> tuple[float, list[str]]:
    divs = [d.dividends_paid for d in accs]

    # Count consecutive years (most recent first) where div >= prior year
    streak = 0
    for i in range(len(divs) - 1, 0, -1):
        if divs[i] >= divs[i - 1] and divs[i] > 0:
            streak += 1
        else:
            break

    if streak >= 10:
        score = 1.0
    elif streak >= 7:
        score = 0.75
    elif streak >= 4:
        score = 0.55
    elif streak >= 2:
        score = 0.35
    elif streak >= 1:
        score = 0.15
    else:
        score = 0.0

    track = (
        "aristocrat-level"
        if streak >= 10
        else "strong track record"
        if streak >= 7
        else "growing dividend"
        if streak >= 4
        else "short track record"
        if streak >= 2
        else "limited history"
        if streak >= 1
        else "dividend cut or freeze detected"
    )
    return score, [
        f"Dividend growth track: {streak} consecutive years maintained  [{track}]",
    ]


# ── Result types ───────────────────────────────────────────────────────────────


@dataclass
class DividendComponents:
    fcf_payout_ratio: float  # 0–1
    earnings_payout_ratio: float  # 0–1
    balance_sheet_strength: float  # 0–1
    dividend_growth_track: float  # 0–1


@dataclass
class DividendSafetyScore:
    """Result of the Dividend Safety Score computation."""

    score: int  # 0–100
    rating: str  # 'safe' | 'adequate' | 'risky' | 'danger' | 'non-payer'
    components: DividendComponents
    years_analyzed: int
    is_dividend_payer: bool
    evidence: list[str] = field(default_factory=list)

    @property
    def interpretation(self) -> str:
        descs = {
            "safe": "Dividend is well-covered by free cash flow and earnings, supported "
            "by a strong balance sheet and a consistent growth track record.",
            "adequate": "Dividend is reasonably covered with moderate payout ratios and "
            "adequate balance sheet headroom.",
            "risky": "Dividend coverage is thin — high payout ratios or elevated leverage "
            "increase the risk of a cut in a downturn.",
            "danger": "Dividend appears unsustainable — payout ratios exceed cash flows "
            "or the balance sheet is severely strained.",
            "non-payer": "Company does not currently pay a dividend. "
            "Neutral score assigned for composite scoring purposes.",
        }
        return (
            f"Dividend Safety Score: {self.score}/100 [{self.rating.upper()}]. "
            f"{descs.get(self.rating, '')}"
        )

    def table(self) -> str:
        w = 52
        sep = "─" * w
        payer_note = "" if self.is_dividend_payer else "  [non-payer]"
        return "\n".join(
            [
                f"Dividend Safety Score: {self.score}/100  [{self.rating.upper()}]{payer_note}",
                sep,
                f"{'Component':<34} {'Score':>7}  {'Weight':>6}",
                sep,
                f"{'FCF Payout Ratio':<34} {self.components.fcf_payout_ratio * 100:>6.0f}%   {'35%':>6}",
                f"{'Earnings Payout Ratio':<34} {self.components.earnings_payout_ratio * 100:>6.0f}%   {'25%':>6}",
                f"{'Balance Sheet Strength':<34} {self.components.balance_sheet_strength * 100:>6.0f}%   {'25%':>6}",
                f"{'Dividend Growth Track':<34} {self.components.dividend_growth_track * 100:>6.0f}%   {'15%':>6}",
                sep,
                f"{'Years of data analyzed':<34} {self.years_analyzed:>7}",
            ]
        )

    def _repr_html_(self) -> str:
        colours = {
            "safe": "#1a7f37",
            "adequate": "#0969da",
            "risky": "#9a6700",
            "danger": "#cf222e",
            "non-payer": "#57606a",
        }
        c = colours.get(self.rating, "#57606a")
        rows = [
            ("FCF Payout Ratio", self.components.fcf_payout_ratio, "35%"),
            ("Earnings Payout Ratio", self.components.earnings_payout_ratio, "25%"),
            ("Balance Sheet Strength", self.components.balance_sheet_strength, "25%"),
            ("Dividend Growth Track", self.components.dividend_growth_track, "15%"),
        ]
        row_html = "".join(
            f"<tr><td>{n}</td><td style='text-align:right'>{v * 100:.0f}%</td>"
            f"<td style='text-align:right;color:#57606a'>{wt}</td></tr>"
            for n, v, wt in rows
        )
        payer_badge = (
            ""
            if self.is_dividend_payer
            else "<span style='margin-left:8px;font-size:0.75em;color:#57606a'>[non-payer]</span>"
        )
        return (
            f"<div style='font-family:monospace;border:1px solid #d0d7de;border-radius:6px;"
            f"padding:16px;max-width:500px'>"
            f"<div style='font-size:1.1em;font-weight:bold;color:{c}'>"
            f"Dividend Safety: {self.score}/100 "
            f"<span style='font-size:0.9em'>[{self.rating.upper()}]</span>{payer_badge}</div>"
            f"<table style='width:100%;border-collapse:collapse;margin-top:12px'>"
            f"<tr style='border-bottom:1px solid #d0d7de'>"
            f"<th style='text-align:left'>Component</th>"
            f"<th style='text-align:right'>Score</th>"
            f"<th style='text-align:right'>Weight</th></tr>"
            f"{row_html}</table>"
            f"<div style='margin-top:8px;color:#57606a;font-size:0.85em'>"
            f"{self.years_analyzed} years analyzed</div>"
            f"</div>"
        )

    def to_dict(self) -> dict:
        return {
            "score": self.score,
            "rating": self.rating,
            "components": {
                "fcf_payout_ratio": round(self.components.fcf_payout_ratio, 4),
                "earnings_payout_ratio": round(self.components.earnings_payout_ratio, 4),
                "balance_sheet_strength": round(self.components.balance_sheet_strength, 4),
                "dividend_growth_track": round(self.components.dividend_growth_track, 4),
            },
            "years_analyzed": self.years_analyzed,
            "is_dividend_payer": self.is_dividend_payer,
            "evidence": self.evidence,
            "interpretation": self.interpretation,
        }


# ── Public API ─────────────────────────────────────────────────────────────────


def dividend_safety_score_from_series(
    annual_data: Sequence[Any],
) -> DividendSafetyScore:
    """Compute Dividend Safety Score from a sequence of annual financial records.

    Parameters
    ----------
    annual_data:
        Sequence of annual records (dicts, dataclasses, or objects with attributes).
        Must be in chronological order (oldest first). Minimum 2 years required.

    Returns
    -------
    DividendSafetyScore
        Dataclass with ``.score`` (0–100), ``.rating``, ``.components``,
        ``.is_dividend_payer``, ``.evidence``, ``.table()``,
        ``._repr_html_()``, and ``.to_dict()``.

    Notes
    -----
    If no dividends are recorded in any year, the score is set to 50 and
    ``rating`` is ``'non-payer'``, ``is_dividend_payer`` is ``False``.
    """
    if len(annual_data) < 2:
        raise ValueError("dividend_safety_score requires at least 2 years of data.")

    accs = [_Acc(d) for d in annual_data]

    # Check if company is a dividend payer
    total_divs = sum(d.dividends_paid for d in accs)
    if total_divs <= 0:
        return DividendSafetyScore(
            score=50,
            rating="non-payer",
            components=DividendComponents(
                fcf_payout_ratio=0.50,
                earnings_payout_ratio=0.50,
                balance_sheet_strength=0.50,
                dividend_growth_track=0.50,
            ),
            years_analyzed=len(accs),
            is_dividend_payer=False,
            evidence=["Company has not paid dividends in the analyzed period."],
        )

    s_fcf, ev_fcf = _score_fcf_payout(accs)
    s_eps, ev_eps = _score_earnings_payout(accs)
    s_bs, ev_bs = _score_balance_sheet(accs)
    s_dgt, ev_dgt = _score_dividend_growth_track(accs)

    raw = 0.35 * s_fcf + 0.25 * s_eps + 0.25 * s_bs + 0.15 * s_dgt
    score = round(_clamp(raw, 0.0, 1.0) * 100)

    rating = (
        "safe"
        if score >= 70
        else "adequate"
        if score >= 45
        else "risky"
        if score >= 20
        else "danger"
    )

    return DividendSafetyScore(
        score=score,
        rating=rating,
        components=DividendComponents(
            fcf_payout_ratio=round(s_fcf, 4),
            earnings_payout_ratio=round(s_eps, 4),
            balance_sheet_strength=round(s_bs, 4),
            dividend_growth_track=round(s_dgt, 4),
        ),
        years_analyzed=len(accs),
        is_dividend_payer=True,
        evidence=ev_fcf + ev_eps + ev_bs + ev_dgt,
    )


def dividend_safety_score(
    ticker: str,
    years: int = 10,
    source: str = "yahoo",
) -> DividendSafetyScore:
    """Fetch data and compute Dividend Safety Score for a ticker.

    Parameters
    ----------
    ticker : str
        Stock ticker symbol (e.g. 'JNJ').
    years : int
        Number of years of historical data to fetch (default 10).
    source : str
        Data source: 'yahoo' (default) or 'edgar'.
    """
    try:
        if source == "edgar":
            from ..fetchers.edgar import fetch_edgar

            raw = fetch_edgar(ticker, num_years=years)
        else:
            from ..fetchers.yahoo import fetch_yahoo

            raw = [fetch_yahoo(ticker)]
    except Exception as exc:
        raise RuntimeError(f"Failed to fetch data for {ticker!r}: {exc}") from exc
    return dividend_safety_score_from_series(raw)


dividend_safety_score.formula = (  # type: ignore[attr-defined]
    "0.35*FCFPayoutRatio + 0.25*EarningsPayoutRatio + 0.25*BalanceSheetStrength + 0.15*DividendGrowthTrack"
)
dividend_safety_score.description = (  # type: ignore[attr-defined]
    "Dividend safety score 0–100. Safe ≥70, Adequate 45–69, Risky 20–44, Danger <20. "
    "Non-payers receive score=50 with rating='non-payer'."
)
